src/main/scala/scorex/core/ModifiersCache.scala

```diff
@@ -0,0 +1,136 @@
+package scorex.core
+
+
+import scorex.core.consensus.HistoryReader
+import scorex.core.utils.ScorexLogging
+import scorex.core.validation.RecoverableModifierError
+
+import scala.collection.mutable
+import scala.util.{Failure, Success}
+
+/**
+  * A cache which is storing persistent modifiers not applied to history yet.
+  * @tparam PMOD - type of a persistent node view modifier (or a family of modifiers).
+  */
+trait ModifiersCache[PMOD <: PersistentNodeViewModifier] {
+  require(maxSize >= 1)
+
+  def size: Int = currentCacheSize
+
+  /**
+    * How many elements are to be stored in the cache
+    */
+  def maxSize: Int
+
+  type H <: HistoryReader[PMOD, _]
+
+  type K = mutable.WrappedArray[Byte]
+  type V = PMOD
+
+  protected var currentCacheSize: Int = 0
+
+  protected val cache = mutable.Map[K, V]()
+
+  protected val rememberedKeys = mutable.HashSet[K]()
+
+  /**
+    * Defines a best (and application-specific) candidate to be applied.
+    * @param history - an interface to history which could be needed to define a candiate
+    * @return - candidate if it is found
+    */
+  def findCandidateKey(history: H): Option[K]
```

Comment:
Maybe the Cache is not the best place to locate this method. Alternatively, this method could be  implemented in the application that uses the cache instead. This would make the cache more agnostic and "ignorant" about application-specific logic, and would allow the cache to just cache. Maybe we could even use a cache library then (e.g. https://github.com/blemale/scaffeine or https://cb372.github.io/scalacache/).

Reply:
Agree with @ceilican.
In addition, looks like it won't be possible to implement `// TODO blaklist peer who sent it` inside a Cache (and it's not a correct place for it)

Reply:
@ceilican @catena2w then the abstract method will be located in NodeViewHolder, which is already overload with diverse functionalities to concretize in an application. On banning the peer, could we use an event bus to send a signal to the network layer? We have work here been abandoned on the way unfortunately.